README
naim(1)              AIM/ICQ/IRC/Lily Communication Client             naim(1)



NAME
       naim - console mode chat client

SYNOPSIS
       naim
       nicq
       nirc [nickname [server]]
       nlily

DESCRIPTION
       naim  is the original ncurses AIM client. It uses the TOC protocol, and
       features many commonly-requested features  found  nowhere  else,  while
       still providing an intuitive chat interface.


OPTIONS
       If  a  `.naimrc' file exists in your home directory, naim executes that
       and ignores all command line options. Otherwise...

       When invoked as `naim', naim will create a connection of  type  AIM/TOC
       and  will  display  some  helpful  information to the screen (but won't
       actually try to sign you on).

       When invoked as `nicq', naim will create a connection of  type  ICQ/TOC
       and  will  display  some  helpful  information to the screen (but won't
       actually try to sign you on).

       When invoked as `nirc', naim will create a connection of type  IRC  and
       will attempt to sign you on using the nickname and server, if provided.

       When invoked  as  `nlily',  naim  will  create  a  connection  of  type
       Lily/SLCP  and will display some helpful information to the screen (but
       won't actually try to sign you on).


ADDITIONAL CONNECTIONS
       Once you have naim started, you can easily  create  additional  connec-
       tions  by  using  the /newconn command. For example, to visit me on the
       EFnet IRC network, you might type:
       /newconn EFnet IRC
       /connect naimuser irc.servercentral.net
       /join #naim
       (note that if you start naim as `nirc' it will run  the  previous  com-
       mands automatically)

       If you wanted to sign on to AIM more than once, you might type:
       /newconn AIM2 AIM
       /connect othername

       If  you start naim as `naim' it will issue `/newconn AIM TOC' automati-
       cally. If you start naim as `nicq' it will  issue  `/newconn  ICQ  TOC'
       automatically.  If  you  start  naim as `nlily' it will issue `/newconn
       Lily SLCP' automatically.  The first argument to /newconn is  the  con-
       nection "name", which can be anything you dream up (it might make sense
       to use your screen name if you are going to be connecting to  the  same
       service  more  than  once),  and  the second argument is the connection
       type. AIM and ICQ both use the same connection type (TOC).

       By default, the Ins and Del keys will switch between  connections.  See
       `/help  keys'  for additional information, including how to change your
       key bindings.


SETTINGS
       Once you have setup naim the way you like it (see `/help settings'  and
       `/help  keys'),  use  the `/save' command to have naim create a .naimrc
       file in your home directory. From then on, naim will  always  start  up
       configured  the  way  it is currently configured, including all of your
       current settings (/set), key bindings  (/bind),  and  open  connections
       (/newconn). Feel free to load the generated .naimrc file in a text edi-
       tor for further tweaking.


ONLINE HELP
       Online help is available by typing /help at the console.


EXITING
       Use the /quit command.


AUTHOR
       Daniel Reed <n@ml.org>, with notable contributions  from  Ian  Gulliver
       <ian@penguinhosting.net> and Joshua Wise <joshua@joshuawise.com>.


UPDATES
       The   latest   version   of   naim   will   always  be  available  from
       http://naim.n.ml.org/ . If you ever use naim, you are strongly  encour-
       aged to subscribe to the naim-announce mailing list. This is a low-vol-
       ume mailing list used exclusively to announce new releases of  naim  or
       changes  in the AIM service that affects naim users. To subscribe, sim-
       ply send a blank email to <naim-announce-subscribe@n.ml.org> and  reply
       to the confirmation message you will receive.


BUG REPORTS
       Before reporting any bugs, please review Simon Tatham (of PuTTY)'s won-
       derful essay, How to Report Bugs Effectively at http://www.chiark.gree-
       nend.org.uk/~sgtatham/bugs.html.

       To  report  issues  with  the front end (things having to do with buddy
       windows, colors, .naimrc, and anything else  "user  oriented")  contact
       the naim maintainer, Daniel Reed <n@ml.org>.

       To  report issues with the protocol library (problems connecting, error
       messages while connected, and anything else not a  part  of  the  front
       end)  contact the FireTalk maintainer, now also Daniel Reed <n@ml.org>.

       To  report  issues  with  the  FreeBSD  port,  contact  Ryan  T.   Dean
       <rtdean@cytherianage.net>.


FEATURE REQUESTS
       If  you  would  like  to  suggest  new features or significant behavior
       changes, subscribe to the naim-users mailing list and suggest it there.
       I  get  a  lot  of mail every day, and non-bug reports get a fairly low
       priority. The best way to make sure something you want  is  implemented
       is  to  post  it  to naim-users whenever I ask for them on the list. To
       subscribe, send a blank email  to  <naim-users-subscribe@n.ml.org>  and
       reply to the confirmation message you will receive.


PATCHES
       If you would like to participate to naim's development in a more active
       role, feel free to submit patches either to myself or to the naim-users
       mailing  list.  Patches in context output format are prefered. In order
       to work on naim, you may wish to extract naim twice, and  perform  your
       builds in a third directory, as in:
        tar -zxvf naim-0.11.8.tar.bz2
        mv naim-0.11.8 naim-0.11.8,original
        tar -zxvf naim-0.11.8.tar.bz2
        mkdir naim-0.11.8-objdir
        cd naim-0.11.8-objdir
        ../naim-0.11.8/configure --prefix=/usr

       Then  you may modify the files in ../naim-0.11.8, compile, test, modify
       again, recompile, test, etc. until your are satisfied, then:
        cd ..
        diff -rcN naim-0.11.8,original naim-0.11.8 > \
              naim-0.11.8.feature1.patch

       And submit  `naim-0.11.8.feature1.patch'.  Do  not  gzip  or  otherwise
       encode patches, so they may be reviewed from within a mail client.




naim-0.11.8                       2005-10-26                           naim(1)


ChangeLog
Summary of changes from 0.11.4.13 to 0.11.5 [2002-12-10]
============================================
[pkg fea] New Spanish README.es.
[UI  fea] AUTOCRYPT, naim will automatically encrypt messages sent to other naim users.
[UI  fea] /save will now store [practically] *all* configuration settings.
[UI  fea] Newer, more transparent client to client protocol.
[UI  fea] Ability to control what types of events cause a window to become yellow, via the $chatter variable:
	# Private messages: +1
	# You kicked, oped, or deoped in a chat: +2
	# Anyone else kicked or deoped in a chat: +4
	# Public messages to a chat: +8
	# Topic changes, or anyone oped in a chat: +16
	# Anyone joining, parting, changing nicknames, or quitting altogether: +32
[UI  fea] The nsocket module has been expanded to better illustrate naim's powerful module interface.

Summary of changes from 0.11.5 to 0.11.5.1 [2002-12-21]
============================================
[UI  fea] Links now appear somewhat cleaner, and redundant links, such as ``<a href="http://something/">http://something/</a>'' do not appear as links (to de-WinAIM messages).
[pkg bug] Many autoconf-related cleanups.
[Lil fea] Minor Lily enhancements to address grossly cross-posted messages.

Summary of changes from 0.11.5.1 to 0.11.5.2-2003-01-01-2004
============================================
[UI  FEA] Time zone information is exchanged during autopeer negotiation.
[UI  fea] Small fixes in the original HTML parser, new HTML parser included (but not yet used).
[UI  fea] The time stamps prepended to messages are now based on the format specified in $timeformat, rather than a hardcoded "[%H:%M:%S]&nbsp;".
[pkg bug] The new user course has been temporarily removed until it can be completed.
[UI  FEA] The window list window now uses ASCII line drawing symbols to promote state.
[UI  fea] Primitive menuing added in (accessed with Escape). Incomplete.
[UI  bug] Apparently /alias actually works; added autohelp message for it.
[UI  fea] Buddy Tab completion will match real names as well as screen names (so "/j<Tab>dani<Tab>" will expand to "/jump RPIDan ", since my real name is set to "Daniel Reed").
[UI  BUG] High ASCII input will no longer cause the input handler to reset itself; it is simply ignored now.
[UI  bug] Home and End will work to switch buddy windows even inside the status window (F1).
[UI  fea] If you are in paste mode, the status bar will identify that fact, and tell you how to get out of it.

Summary of changes from 0.11.5.2-2003-01-01-2004 to 0.11.5.2-2003-01-03-0149
============================================
[UI  bug] /set completion fixed.
[UI  bug] KEY_RESIZE handler corrected.

Summary of changes from 0.11.5.2-2003-01-03-0149 to 0.11.5.2-2003-01-04-2111
============================================
[AIM FEA] Profiles that are too long to meet the server's size limits will be truncated to fit, and a warning will be sent to the user.
[UI  bug] Small bugs fixed in the /quit cleanup code (so /disconnect works again).
[UI  fea] Tab complete now fills in connection windows if no commands match. /set completion makes Even More Sense(TM) now.

Summary of changes from 0.11.5.2-2003-01-04-2111 to 0.11.5.2-2003-01-06-1445
============================================
[UI  fea] When naim joins a chat, people already in the chat will be displayed in one message (rather than one message per person).
[Lil bug] Fixed a seg fault if you weren't in any chats on startup.
[UI  bug] Screen names exactly as wide as the window list window would have the last character replaced with a '>'. Fixed.
[UI  fea] Up-arrow completion now has the ghost effect /-commands have as of 0.11.5.
[UI  FEA] New $awaylog. If set, all messages received while /away will be mirrored in an :AWAYLOG window (for people who detach for days and lose messages from the 500 line scrollback limit).

Summary of changes from 0.11.5.2-2003-01-06-1445 to 0.11.5.2-2003-01-07-0336
============================================
[Yah fea] Preliminary Yahoo! Messenger support begun.
[pkg fea] New --enable-firetalk-msn and --enable-firetalk-yahoo, both default to disabled. Once these protocol drivers are able to successfully sign on and send/receive IMs, they will always be built.
[UI  fea] :-windows (such as :RAW, :AWAYLOG, :REVIEW) are now fully treated as special cases in naim, and naim won't attempt to join them on signon.
[AIM bug] Chat name normalization bug squashed.
[UI  FEA] New $autosave. If set (default currently off), naim will /save when the user /quits.
[UI  fea] New /winlist command. Allows the user to set the state of the window list window, a la F4. /save uses this command to restore your F4 state on start-up.
[UI  FEA] People who are /ignored will be ignored in group chats as well. This is mainly so I can /ignore stupid bots in channels I frequent, but others have requested this in the past.
[UI  fea] $awaylog has been expanded: A value of 1 logs only private messages received, while a value greater than 1 logs private messages sent and chat messages.

Summary of changes from 0.11.5.2-2003-01-07-0336 to 0.11.5.2-2003-01-07-0515
============================================
[UI  fea] Misc. UI cleanups. Due to popular request, names are now colourized in the :AWAYLOG window.

Summary of changes from 0.11.5.2-2003-01-07-0515 to 0.11.5.2-2003-01-08-1249
============================================
[Lil fea] The RPI Lily community dislikes naim's UI, and fears its proliferation will be the downfall of Lily. In any event, naim now combines multiple /joins. Emote discussions are now treated properly (messages are prepended with ", emotes are not prepended with /me).
[UI  bug] -01-04-2111 lied. /quit and /disconnect *really* work correctly now.

Summary of changes from 0.11.5.2-2003-01-08-1249 to 0.11.5.2-2003-01-10-0308
============================================
[pkg fea] New --disable-nirc-manlink. If specified, the nirc man page symlink will not be installed (for FreeBSD).

src/rc.c
/*  _ __   __ _ ___ __  __
** | '_ \ / _` |_ _|  \/  | naim
** | | | | (_| || || |\/| | Copyright 1998-2004 Daniel Reed <n@ml.org>
** |_| |_|\__,_|___|_|  |_| ncurses-based chat client
*/
#include <naim/naim.h>
#include <fcntl.h>

#include "naim-int.h"

extern conn_t	*curconn;
extern const char	*home;
extern time_t	now;

buddylist_t
	*raddbuddy(conn_t *conn, const char *screenname, const char *group,
	const char *notes) {
	buddylist_t	*buddy = NULL;

	assert(conn != NULL);
	assert(screenname != NULL);
	assert(group != NULL);

	for (buddy = conn->buddyar; buddy != NULL; buddy = buddy->next)
		if (firetalk_compare_nicks(conn->conn, screenname, buddy->_account) == FE_SUCCESS)
			return(buddy);

	buddy = calloc(1, sizeof(buddylist_t));
	assert(buddy != NULL);
	STRREPLACE(buddy->_account, screenname);
	STRREPLACE(buddy->_group, group);
	if (notes != NULL)
		STRREPLACE(buddy->_name, notes);
	else
		FREESTR(buddy->_name);

	if ((conn->buddyar == NULL) || (aimcmp(screenname, conn->buddyar->_account) <= 0)) {
		buddy->next = conn->buddyar;
		conn->buddyar = buddy;
	} else {
		buddylist_t	*btmp = conn->buddyar;

		while ((btmp->next != NULL) && (aimcmp(screenname, btmp->next->_account) > 0))
			btmp = btmp->next;
		buddy->next = btmp->next;
		btmp->next = buddy;
	}

	buddy->offline = 1;
	buddy->isaway = buddy->isidle = 0;

	return(buddy);
}

void	do_delbuddy(buddylist_t *b) {
	FREESTR(b->_account);
	FREESTR(b->_group);
	FREESTR(b->_name);
	FREESTR(b->crypt);
	FREESTR(b->tzname);
	FREESTR(b->tag);
	FREESTR(b->caps);
	free(b);
}

void	rdelbuddy(conn_t *conn, const char *screenname) {
	buddylist_t	*buddy = NULL;

	assert(conn != NULL);
	assert(screenname != NULL);

	assert(bgetwin(conn, screenname, BUDDY) == NULL);

	if ((buddy = conn->buddyar) == NULL)
		return;

	if (firetalk_compare_nicks(conn->conn, screenname, buddy->_account) == FE_SUCCESS) {
		if ((conn->online > 0) && (buddy->peer > 0))
			firetalk_subcode_send_request(conn->conn, screenname, "AUTOPEER", "-AUTOPEER");
		buddy->peer = 0;
		conn->buddyar = buddy->next;
		do_delbuddy(buddy);
		return;
	}
	while (buddy->next != NULL) {
		if (firetalk_compare_nicks(conn->conn, screenname, buddy->next->_account) == FE_SUCCESS) {
			buddylist_t	*b = buddy->next;

			if ((conn->online > 0) && (b->peer > 0))
				firetalk_subcode_send_request(conn->conn, screenname, "AUTOPEER", "-AUTOPEER");
			b->peer = 0;
			buddy->next = buddy->next->next;
			do_delbuddy(b);
			return;
		}
		buddy = buddy->next;
	}
}

void	raddidiot(conn_t *conn, const char *screenname, const char *notes) {
	ignorelist_t	*idiot = NULL;

	assert(conn != NULL);
	assert(screenname != NULL);
	idiot = conn->idiotar;
	while (idiot != NULL) {
		if (firetalk_compare_nicks(conn->conn, screenname, idiot->screenname) == FE_SUCCESS)
			break;
		idiot = idiot->next;
	}
	if (idiot == NULL) {
		idiot = calloc(1, sizeof(ignorelist_t));
		assert(idiot != NULL);
		idiot->next = conn->idiotar;
		conn->idiotar = idiot;
	}
	STRREPLACE(idiot->screenname, screenname);
	if (notes != NULL)
		STRREPLACE(idiot->notes, notes);
	else
		STRREPLACE(idiot->notes, "There is no reason, you're just"
			" ignored!");
}

void	rdelidiot(conn_t *conn, const char *screenname) {
	ignorelist_t	*idiot = NULL;

	assert(conn != NULL);
	assert(screenname != NULL);

	if ((idiot = conn->idiotar) == NULL)
		return;
	if (firetalk_compare_nicks(conn->conn, screenname, idiot->screenname) == FE_SUCCESS) {
		conn->idiotar = idiot->next;
		free(idiot->screenname);
		idiot->screenname = NULL;
		free(idiot->notes);
		idiot->notes = NULL;
		free(idiot);
		return;
	}
	while (idiot->next != NULL) {
		if (firetalk_compare_nicks(conn->conn, screenname, idiot->next->screenname) == FE_SUCCESS) {
			ignorelist_t	*b = idiot->next;

			idiot->next = idiot->next->next;
			free(b->screenname);
			b->screenname = NULL;
			free(b->notes);
			b->notes = NULL;
			free(b);
			return;
		}
		idiot = idiot->next;
	}
}

buddylist_t
	*rgetlist(conn_t *conn, const char *screenname) {
	buddylist_t	*blist = NULL;

	assert(conn != NULL);
	assert(screenname != NULL);
	if ((blist = conn->buddyar) != NULL)
		do {
			if (firetalk_compare_nicks(conn->conn, blist->_account, screenname) == FE_SUCCESS)
				return(blist);
		} while ((blist = blist->next) != NULL);
	return(NULL);
}

const char
	*buddy_tabcomplete(conn_t *const conn, const char *start, const char *buf, const int bufloc, int *const match, const char **desc) {
	static char
		str[1024];
	buddylist_t	*blist = NULL;
	size_t	slen;

	assert(start != NULL);
	slen = strlen(start);
	if ((blist = conn->buddyar) != NULL)
		do {
			if (aimncmp(blist->_account, start, slen) == 0) {
				const char
					*name = blist->_account;
				int	j;

				if (match != NULL)
					*match = bufloc - (start-buf);
				if (desc != NULL)
					*desc = blist->_name;
				for (j = 0; (j < sizeof(str)-1) && (*name != 0); j++) {
					while (*name == ' ')
						name++;
					str[j] = *(name++);
				}
				str[j] = 0;
				return(str);
			}
		} while ((blist = blist->next) != NULL);
	return(NULL);
}

#if 0
static int
	naim_eval(const char *str) {
	assert(str != NULL);
	if (*str == 0)
		return(0);
	if (strncasecmp(str, "LINES", 5) == 0) {
		str += 5;
		while ((*str != 0) && isspace(*str))
			str++;
		if (*str == 0)
			return(LINES);
		if (*str == '-') {
			str++;
			while ((*str != 0) && isspace(*str))
				str++;
			return(LINES - atoi(str));
		}
		while ((*str != 0) && isspace(*str))
			str++;
		return(atoi(str));
	}
	if (strncasecmp(str, "COLS", 4) == 0) {
		str += 4;
		while ((*str != 0) && isspace(*str))
			str++;
		if (*str == 0)
			return(COLS);
		if (*str == '-') {
			str++;
			while ((*str != 0) && isspace(*str))
				str++;
			return(COLS - atoi(str));
		}
		while ((*str != 0) && isspace(*str))
			str++;
		return(atoi(str));
	}
	return(atoi(str));
}
#endif

int	rc_resize(faimconf_t *conf) {
	static int
		lastCOLS = 0,
		lastLINES = 0;

	if ((lastCOLS == COLS) && (lastLINES == LINES))
		return(0);
	lastCOLS = COLS;
	lastLINES = LINES;

	conf->wstatus.widthx = COLS;
	conf->wstatus.startx = 0;
	conf->wstatus.widthy = LINES-2;
	conf->wstatus.starty = 0;
	conf->wstatus.pady = secs_getvar_int("scrollback");

	conf->winput.widthx = COLS;
	conf->winput.startx = 0;
	conf->winput.widthy = 1;
	conf->winput.starty = LINES - conf->winput.widthy;
	conf->winput.pady = 0;

	conf->winfo.widthx = COLS;
	conf->winfo.startx = 0;
	conf->winfo.widthy = 1;
	conf->winfo.starty = LINES-2;
	conf->winfo.pady = 0;

	conf->waway.widthx = 44;
	conf->waway.startx = COLS/2 - (conf->waway.widthx)/2;
	conf->waway.widthy = 4;
	conf->waway.starty = LINES/2 - 1 - (conf->waway.widthy)/2;
	conf->waway.pady = 0;

	return(1);
}

rc_var_s_t
	rc_var_s_ar[] = {
	{ "nameformat",		"$user_name_account$user_name_ifwarn",
								"window list name format for unnamed users" },
	{ "nameformat_named",	"$user_name_account$user_name_ifwarn ($user_name_name)",
								"window list name format for named buddies" },
	{ "awaymsg",		"I am currently away...",	"auto-response sent when you are away" },
	{ "autoawaymsg",	"Must have stepped out...",	"auto-response used if you idle away (see $autoaway)" },
	{ "autozone",		NULL,				"time zone to send to peers" },
	{ "logdir",		".naimlog/$conn/$cur",		"base directory to store persistent window scrollback" },
	{ "statusbar",		" %I:%M%p $ifoper$SN$ifwarn$ifaway$ifquery$ifchat+$ifpending$iftransfer [$conn $online]$iflag$ifidle naim ",
								"string to use for status bar, replace $if* with $statusbar_*" },
	{ "statusbar_away",	" (away)",			"display if you are away" },
	{ "statusbar_idle",	" [Idle $idle]",		"display if you haven't said anything for 10 minutes" },
	{ "statusbar_lag",	" [Lag $lag]",			"display if there is network delay between you and the server" },
	{ "statusbar_oper",	"@",				"display if the current window is a chat, and you are a channel operator" },
	{ "statusbar_warn",	"/$warnval%",			"if your warning level is above 0%" },
	{ "statusbar_chat",	" on $cur$iftopic",		"display if the current window is a chat" },
	{ "statusbar_crypt",	"ENCRYPTED ",			"display if the current window is a query, and query is automatically encrypted" },
	{ "statusbar_typing",	" TYPING",			"if the current window is a query, and the other person is typing to you" },
	{ "statusbar_tzname",	" <$tzname>",			"display if the current window is a query, and you know the buddy's time zone" },
	{ "statusbar_query",	" [$ifcryptQuery: $cur$iftopic$iftzname]$iftyping",
								"display if the current window is a query" },
	{ "timeformat",		"[%H:%M:%S]&nbsp;",		"strftime format prepended to all messages" },
	{ "im_prefix",		"",				"string added to the beginning of every IM sent" },
	{ "im_suffix",		"",				"string added to the end of every IM sent" },
};
const int
	rc_var_s_c = sizeof(rc_var_s_ar)/sizeof(*rc_var_s_ar);

rc_var_i_t
	rc_var_i_ar[] = {
	{ "chatter",		(IM_MESSAGE | CH_ATTACKED | CH_ATTACKS | CH_MESSAGE),
					"determines which events trigger 'updated' status (yellow)" },
	{ "awaylog",		0,	"1 = log IM's to an :AWAYLOG window when away, 2 = IM's and chat messages" },
	{ "echostyle",		(ALSO_STATUS),
					"determines how /echo behaves" },
	{ "winlistheight",	90,	"percentage of the screen height to use for the window list window" },
	{ "winlistchars",	16,	"number of characters from each window name to display in the window list" },
	{ "beeponim",		1,	"beep whenever an IM is received (0 = never, 1 = if not away, 2 = always)" },
	{ "beeponsignon",	0,	"beep whenever a buddy signs on or off (0 = never, 1 = if not away, 2 = always)" },
	{ "autoaway",		20,	"if non-zero, maximum idle time before automatically marking away" },
	{ "autoreply",		10,	"number of minutes to wait before sending a second auto-response" },
	{ "tprint",		30,	"if non-zero, a timestamp is displayed periodically in all query windows" },
	{ "logtprint",		2,	"1 = log daily timestamps, 2 = log all timestamps" },
	{ "autohide",		3,	"number of seconds to display the status and window list windows" },
	{ "idletime",		0,	"number of minutes you have been idle" },
	{ "scrollback",		500,	"number of lines in the scrollback buffer" },
	{ "autoclose",		15,	"number of minutes to keep a non-permanent (autobuddy'd) window open after the user signs off" },
	{ "autosort",		2,	"string sorting type (0 = none, 1 = by name, 2 = by group)" },
#ifdef ENABLE_DNSUPDATE
	{ "updatecheck",	4*60,	"number of minutes to wait to check for new versions of naim" },
#endif
};
const int
	rc_var_i_c = sizeof(rc_var_i_ar)/sizeof(*rc_var_i_ar);

rc_var_i_t
	rc_var_b_ar[] = {
	{ "autounaway",		0,	"automatically unmark away status whenever an IM is sent" },
	{ "autoidle",		1,	"automatically increment $idletime by 1 every minute" },
	{ "autounidle",		1,	"reset $idletime to 0 whenever an IM is sent" },
	{ "lagcheck",		1,	"periodically determine delay to server" },
	{ "autoquery",		1,	"whenever a buddy signs on, a temporary window will be opened for them" },
	{ "autopaste",		1,	"enable logic to detect accidental pastes without first entering paste mode" },
	{ "autobuddy",		1,	"whenever a non-buddy IM's you, a window will be opened for them" },
	{ "autopeer",		1,	"automatically attempt to negotiate client capabilities" },
	{ "autopeerverbose",	0,	"notify you as client capabilities are negotiated" },
	{ "autocrypt",		1,	"use autopeer to negotiate peer to peer IM encryption" },
	{ "autosave",		0,	"automatically update .naimrc on /quit" },
	{ "ctcpverbose",	0,	"notify you of unknown CTCP requests" },
	{ "color",		1,	"honor <font color=...> tags in IM's" },
	{ "autoreconnect",	0,	"automatically reconnect if disconnected; you may need to /set an appropriate password variable (such as $AIM:password)" },
	{ "autonames",		1,	"automatically /names on /join" },
	{ "showraw",		0,	"display special server messages in a window named :RAW" },
#ifdef ENABLE_FORCEASCII
	{ "forceascii",		0,	"always use ASCII linedraw chars" },
#endif
};
const int
	rc_var_b_c = sizeof(rc_var_b_ar)/sizeof(*rc_var_b_ar);


#define timeoff		(-100*timezone/(60*60))
#define abstimeoff	((timeoff < 0)?-timeoff:timeoff)
#define chrtimeoff	((timeoff < 0)?'-':'+')

void	rc_initdefs(faimconf_t *conf) {
	static char
		autozone[1024];
	int	i;

	memset(conf, 0, sizeof(*conf));

#ifdef HAVE_STRUCT_TM_TM_ZONE
	{
		struct tm
			*tmptr;

		if ((tmptr = localtime(&now)) != NULL) {
			if (strchr(tmptr->tm_zone, ' ') == NULL)
				snprintf(autozone, sizeof(autozone), "%s%c%04li", 
					tmptr->tm_zone, chrtimeoff, abstimeoff);
			else
				*autozone = 0;
		} else
			*autozone = 0;
	}
#elif HAVE_TZNAME
	{
		extern char
			*tzname[2];

		localtime(&now);
		if (strchr(tzname[0], ' ') == NULL)
			snprintf(autozone, sizeof(autozone), "%s%c%04li", tzname[0], chrtimeoff, 
				abstimeoff);
		else
			*autozone = 0;
	}
#else
	*autozone = 0;
#endif
	for (i = 0; i < rc_var_s_c; i++)
		if (strcmp(rc_var_s_ar[i].var, "autozone") == 0) {
			rc_var_s_ar[i].val = autozone;
			break;
		}

#ifdef DEBUG_ECHO
	for (i = 0; i < rc_var_b_c; i++)
		if (strcmp(rc_var_b_ar[i].var, "lagcheck") == 0) {
			rc_var_b_ar[i].val = 0;
			break;
		}
	for (i = 0; i < rc_var_b_c; i++)
		if (strcmp(rc_var_b_ar[i].var, "showraw") == 0) {
			rc_var_b_ar[i].val = 1;
			break;
		}
	for (i = 0; i < rc_var_i_c; i++)
		if (strcmp(rc_var_i_ar[i].var, "autoaway") == 0) {
			rc_var_i_ar[i].val = 0;
			break;
		}
#endif

	if (secs_getvar("lag") == NULL)
		secs_setvar("lag", "0s");
	if (secs_getvar("SN") == NULL)
		secs_setvar("SN", "- Type /connect \"screen name\" (include the quotes)");
	if (secs_getvar("lag") == NULL)
		secs_setvar("online", "(offline)");

	for (i = 0; i < rc_var_s_c; i++)
		if (secs_getvar(rc_var_s_ar[i].var) == NULL)
			secs_setvar(rc_var_s_ar[i].var, rc_var_s_ar[i].val);

	for (i = 0; i < rc_var_i_c; i++)
		if (secs_getvar(rc_var_i_ar[i].var) == NULL)
			secs_makevar_int(rc_var_i_ar[i].var, rc_var_i_ar[i].val, 'I', NULL);

	for (i = 0; i < rc_var_b_c; i++)
		if (secs_getvar(rc_var_b_ar[i].var) == NULL)
			secs_makevar_int(rc_var_b_ar[i].var, rc_var_b_ar[i].val, 'B', NULL);

	conf->f[cEVENT] = 3;
	conf->f[cEVENT_ALT] = 2;
	conf->f[cTEXT] = 7;
	conf->f[cSELF] = 1;
	conf->f[cBUDDY] = 6;
	conf->f[cBUDDY_WAITING] = 3;
	conf->f[cBUDDY_ADDRESSED] = 3;
	conf->f[cBUDDY_IDLE] = 4;
	conf->f[cBUDDY_AWAY] = 2;
	conf->f[cBUDDY_OFFLINE] = 1;
	conf->f[cBUDDY_QUEUED] = 5;
	conf->f[cBUDDY_TAGGED] = 4;

	conf->b[cINPUT] = 0;
	conf->b[cWINLIST] = 6;
	conf->b[cWINLISTHIGHLIGHT] = 0;
	conf->b[cCONN] = 4;
	conf->b[cIMWIN] = 0;
	conf->b[cSTATUSBAR] = 7;

	rc_resize(conf);
}

int	naim_read_config(const char *faimrcfile) {
	FILE	*faimrc = NULL;
	char	buf[1024];
	int	linenum = 0;

	if (strlen(faimrcfile) < 1)
		return(0);
	if (isspace(faimrcfile[strlen(faimrcfile)-1])) {
		strncpy(buf, faimrcfile, sizeof(buf)-1);
		buf[sizeof(buf)-1] = 0;
		while (isspace(buf[strlen(buf)-1]))
			buf[strlen(buf)-1] = 0;
		faimrcfile = buf;
	}
	if ((faimrc = fopen(faimrcfile, "r")) == NULL) {
		echof(curconn, NULL, "Unable to open %s.\n",
			faimrcfile);
		return(0);
	}
	buf[sizeof(buf)-1] = 0;
	while (!feof(faimrc)) {
		fgets(buf, sizeof(buf)-1, faimrc);
		if (feof(faimrc))
			break;
		linenum++;
		if ((*buf == 0) || (*buf == '#'))
			continue;
		while ((*buf != 0) && ((buf[strlen(buf)-1] == '\n')
			|| (buf[strlen(buf)-1] == '\r')))
			buf[strlen(buf)-1] = 0;
		if (*buf == 0)
			continue;
		conio_handlecmd(buf);
	}
	fclose(faimrc);
	return(1);
}
